pkg/controllers/managementuser/resourcequota/resource_quota_sync.go

```diff
@@ -119,69 +118,84 @@ func (c *SyncController) CreateResourceQuota(ns *corev1.Namespace) (runtime.Obje
 		return ns, err
 	}
 
-	projectLimit, _, err := getProjectResourceQuotaLimit(ns, c.ProjectLister)
-	if err != nil {
-		return ns, err
-	}
-
-	var quotaSpec *corev1.ResourceQuotaSpec
-	if projectLimit != nil {
-		quotaSpec, err = c.getNamespaceResourceQuota(ns)
+	var nsResourceQuota v32.NamespaceResourceQuota
+	if projectLimit, _, err := getProjectResourceQuotaLimit(ns, c.ProjectLister); err == nil && projectLimit != nil {
+		nsResourceQuota, err = c.getNamespaceResourceQuota(ns)
 		if err != nil {
 			return ns, err
 		}
 	}
 
-	quotaToUpdate, err := c.getResourceQuotaToUpdate(ns)
+	quotaToUpdate, err := c.getResourceQuotaToUpdate(ns, &nsResourceQuota)
 	if err != nil {
 		return ns, err
 	}
 
+	var updated *corev1.Namespace
+	newQuotaSpec, err := convertResourceLimitResourceQuotaSpec(&nsResourceQuota.Limit)
+	if err != nil {
+		return nil, err
+	}
+
 	operation := "none"
-	if existing == nil {
-		if quotaSpec != nil {
-			operation = "create"
-		}
+	if existing == nil && newQuotaSpec != nil {
+		operation = "create"
 	} else {
-		if quotaSpec == nil {
+		if newQuotaSpec == nil {
 			operation = "delete"
-		} else if quotaToUpdate != "" || !apiequality.Semantic.DeepEqual(existing.Spec.Hard, quotaSpec.Hard) {
+		} else if quotaToUpdate != nil || !apiequality.Semantic.DeepEqual(existing.Spec.Hard, newQuotaSpec.Hard) {
 			operation = "update"
 		}
 	}
 
-	var updated *corev1.Namespace
-	var isFit bool
+	var operationErr error
 	switch operation {
 	case "create":
-		isFit, updated, err = c.validateAndSetNamespaceQuota(ns, quotaToUpdate)
+		isFit, updated, exceeded, err := c.validateAndSetNamespaceQuota(ns, quotaToUpdate)
 		if err != nil {
 			return updated, err
 		}
 		if !isFit {
-			// create default "all 0" resource quota
-			quotaSpec, err = getDefaultQuotaSpec()
+			// Create a quota with zeros only for overused resources.
+			limit, err := convertResourceListToLimit(exceeded)
+			if err != nil {
+				return updated, err
+			}
+
+			limit, err = zeroOutResourceQuotaLimit(limit)
```

Comment:
Given that the `zeroOutResourceQuotaLimit` function is only used here, it might be more helpful to augment it to handle most of the logic that fits within this if statement. 

For example, that function could be changed to take in a limit (i.e. the limit which had excesses) and a list of resource which exceeded the limits (the exceeded values). It could then overwrite the exceeded values and return a properly parsed resource. This could also have the benefit of allowing us to not use completeQuota - which I think is a bit of a dangerous method given that it combines these two values in a very particular way.

Reply:
Resolving - changes made.